File: scripts/ibgraphfabric.pl
#!/usr/bin/perl
#
# Copyright (c) 2008 Lawrence Livermore National Security
#
# Produced at Lawrence Livermore National Laboratory.
# Written by Ira Weiny <weiny2@llnl.gov>.
# UCRL-CODE-235440
#
# This software is available to you under a choice of one of two
# licenses.  You may choose to be licensed under the terms of the GNU
# General Public License (GPL) Version 2, available from the file
# COPYING in the main directory of this source tree, or the
# OpenIB.org BSD license below:
#
#     Redistribution and use in source and binary forms, with or
#     without modification, are permitted provided that the following
#     conditions are met:
#
#      - Redistributions of source code must retain the above
#        copyright notice, this list of conditions and the following
#        disclaimer.
#
#      - Redistributions in binary form must reproduce the above
#        copyright notice, this list of conditions and the following
#        disclaimer in the documentation and/or other materials
#        provided with the distribution.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
# BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#

use strict;

use Getopt::Std;
use IBswcountlimits;
use Hostlist;

sub usage_and_exit
{
	my $prog = $_[0];
	print
"Usage: $prog [-hRe -S <guid> -C <ca_name> -P <ca_port> -r <string[,string]> ]\n";
	print
"   Report link speed and connection for each port of each switch which is active\n";
	print "   -h This help message\n";
	print
"   -R Recalculate cached information (Default is to reuse cached output)\n";
	print
"   -S <guid> generate for the switch specified by guid and its immediate links\n";
	print "   -C <ca_name> use selected Channel Adaptor name for queries\n";
	print "   -P <ca_port> use selected channel adaptor port for queries\n";
	print "   -r <string[,string]> regex's to search for root switch's\n";
	print
"   -e combine edges between nodes into one edge with an associated weight\n";
	print "   -M <node[,node]> plot routes from node to node in red\n";
	print "   -i <string[,string]> ignore nodes which match regex's given\n";
	print "   -g <cluster_name> DON'T print GUID's on nodes.\n";
	exit 0;
}

my $argv0             = `basename $0`;
my $regenerate_cache  = undef;
my $single_switch     = undef;
my $combine_edges     = undef;
my $ca_name           = "";
my $ca_port           = "";
my $dot_mode          = "1";
my @root_sw_regex     = ();
my @ignore_node_regex = ();
my $plot_nodes        = undef;
my @hosts             = ();
my $print_guids       = "yes";
chomp $argv0;

if (!getopts("hRS:C:P:r:eM:i:g:")) { usage_and_exit $argv0; }
if (defined $Getopt::Std::opt_h)   { usage_and_exit $argv0; }
if (defined $Getopt::Std::opt_R) { $regenerate_cache = $Getopt::Std::opt_R; }
if (defined $Getopt::Std::opt_S) { $single_switch    = $Getopt::Std::opt_S; }
if (defined $Getopt::Std::opt_C) { $ca_name          = $Getopt::Std::opt_C; }
if (defined $Getopt::Std::opt_P) { $ca_port          = $Getopt::Std::opt_P; }
if (defined $Getopt::Std::opt_r) {
	@root_sw_regex = split(",", $Getopt::Std::opt_r);
}
if (defined $Getopt::Std::opt_i) {
	@ignore_node_regex = split(",", $Getopt::Std::opt_i);
}
if (defined $Getopt::Std::opt_g) {
	$print_guids = $Getopt::Std::opt_g;
}
if (defined $Getopt::Std::opt_e) { $combine_edges = $Getopt::Std::opt_e; }
if (defined $Getopt::Std::opt_M) { $plot_nodes    = $Getopt::Std::opt_M; }

my $extra_smpquery_params = get_ca_name_port_param_string($ca_name, $ca_port);

# given a swguid and lid [$routing{$swguid}{$lid}] value is the port to route out
my %routing = undef;
# given a node name, map the valid lids for that node.
my %name2lids = undef;
my @lids      = undef;
if (defined $plot_nodes) {
	read_routing_file($regenerate_cache, $ca_name, $ca_port);
	@hosts = Hostlist::expand($plot_nodes);
	foreach my $host (@hosts) {
		#		print("$host -> @lids\n");
		push(@lids, $name2lids{$host});
	}
	#	print ("@lids\n");
	#	foreach my $t (keys %routing) {
	#		foreach my $t2 (keys %{$routing{$t}}) {
	#			print ("$t,$t2 -> $routing{$t}{$t2}\n");
	#		}
	#	}
}

sub main
{
	my %edges =
	  {}; # double hash which is defined by "node:port","node:port" and has a value of "weight"
	my %nodes = undef;    # nodes with desc and guid stored.
	my @roots = ();
	my @leafs = ();
	get_link_ends($regenerate_cache, $ca_name, $ca_port);
	foreach my $switch (sort (keys(%IBswcountlimits::link_ends))) {
		if ($single_switch && $switch ne $single_switch) {
			next;
		}
		my $num_ports = get_num_ports($switch, $ca_name, $ca_port);
		if ($num_ports == 0) {
			printf("ERROR: switch $switch has 0 ports???\n");
		}
		foreach my $port (1 .. $num_ports) {
			my $hr       = $IBswcountlimits::link_ends{$switch}{$port};
			my $psw      = $switch;
			my $rpsw     = $hr->{rem_guid};
			my $node_rec = undef;
			# skip nodes which match the regex's given
			my $ignore = "no";
			foreach my $regex (@ignore_node_regex) {
				if ($hr->{loc_desc} =~ /.*$regex.*/) { $ignore = "yes"; }
				if ($hr->{rem_desc} =~ /.*$regex.*/) { $ignore = "yes"; }
			}
			if ($ignore eq "yes") { next; }
			$psw =~ s/^0(x.*)/$1/;
			if ($rpsw ne "") {
				$rpsw =~ s/^0(x.*)/$1/;

				# store the nodes.
				$node_rec = {
					desc => $hr->{loc_desc},
					guid => $switch,
					lid  => $hr->{loc_sw_lid}
				};
				$nodes{$psw} = $node_rec;
				$node_rec = {
					desc => $hr->{rem_desc},
					guid => $hr->{rem_guid},
					lid  => $hr->{rem_lid}
				};
				$nodes{$rpsw} = $node_rec;

				# define the edges
				my $p  = $port;
				my $rp = $hr->{rem_port};

				# optionally Combine similar links with a weight
				# Do this by making dot think all links are
				# to "port 0"
				if (defined $combine_edges) {
					$p  = 0;
					$rp = 0;
				}
				my $pswnp  = join(":", $psw,  $p);
				my $rpswnp = join(":", $rpsw, $rp);
				if (exists $edges{$pswnp}{$rpswnp}) {
					$edges{$pswnp}{$rpswnp}++;
				} elsif (exists $edges{$rpswnp}{$pswnp}) {
					# do nothing
				} else {
					$edges{$pswnp}{$rpswnp} = 1;
				}
			}
			foreach my $regex (@root_sw_regex) {
				if ($hr->{loc_desc} =~ /.*$regex.*/) { push(@roots, $psw); }
				if ($hr->{rem_desc} =~ /.*$regex.*/) { push(@roots, $rpsw); }
			}
		}
	}

	# print headers
	printf("digraph G {\n");
	printf("   node [shape=record, fontsize=9];\n");
	my $ranksep = 1 + ((keys %nodes) * 0.5 / 10);
	printf("   graph [outputorder=nodesfirst, ranksep=\"%d equally\"];\n",
		$ranksep);

	# sort the nodes according to their number
	my @sorted_keys = sort {
		my $aa = 0;
		my $bb = 0;
		if ($nodes{$a}->{desc} =~ /\D*(\d*)/) {
			$aa = $1;
		}
		if ($nodes{$b}->{desc} =~ /\D*(\d*)/) {
			$bb = $1;
		}
		$aa <=> $bb;
	} keys %nodes;

	# print nodes
	foreach my $key (@sorted_keys) {
		if ("$key" ne "") {
			my $attr = "";
			if ($print_guids ne "yes"
				&& ($nodes{$key}->{desc} =~ /$print_guids.*/))
			{
				$attr = sprintf("label=\"%s\"", $nodes{$key}->{desc});
			} else {
				$attr = sprintf(
					"label=\"%s\\nG: %s\"",
					$nodes{$key}->{desc},
					$nodes{$key}->{guid}
				);
			}
			my @tmp = ($nodes{$key}->{desc});
			if (Hostlist::within(\@tmp, \@hosts)) {
				$attr = "$attr, color=\"red\"";
			}
			printf("   %18s [%s];\n", $key, $attr);
		}
	}

	# print edges
	foreach my $key1 (keys %edges) {
		foreach my $key2 (keys %{$edges{$key1}}) {
			my ($nd,  $p)  = split(":", $key1);
			my ($rnd, $rp) = split(":", $key2);
			my $attr = "";
			if ($p != 0 || $rp != 0) {
				# if the user did not combine the links print the ports
				# for the edge at the end of the link
				$attr = sprintf("taillabel=\"%d\", headlabel=\"%d\"", $p, $rp);
			} else {
				# combined links just print the "weight"
				$attr = sprintf("label=\"%d\"", $edges{$key1}{$key2});
			}
			# color combined edges blue
			if ($edges{$key1}{$key2} > 1) {
				$attr = "$attr, color=\"blue\"";
			} elsif (defined $plot_nodes) {
				# color requested routes red
				foreach my $lid (@lids) {
					#printf ("0$nd,$lid => %d ?= $p\n", $routing{"0$nd"}{$lid});
					if ($routing{"0$nd"}{$lid} == $p) {
						$attr = "$attr, color=\"red\"";
					} elsif ($routing{"0$rnd"}{$lid} == $rp) {
						$attr = "$attr, color=\"red\"";
					}
				}
			}
			printf("   %18s -> %18s [%s, arrowhead=\"none\"];\n",
				$nd, $rnd, $attr);
		}
	}

	# define roots if there were any
	if (scalar @roots) {
		printf("{rank=min; ");
		foreach my $root (@roots) {
			printf("$root; \n");
		}
		printf("}\n");
	}

	# close dot file
	printf("}\n");
}
main;

####################################
##  dump_lfts data  ################
####################################
#Unicast lids [0x0-0xe5c] of switch Lid 11 guid 0x0008f10400412824 (ISR9024D Voltaire):
#  Lid  Out   Destination
#       Port     Info
#0x0002 007 : (Switch portguid 0x0008f10400412806: 'ISR9024D Voltaire')
#0x0003 003 : (Switch portguid 0x0008f10400411a0e: 'ISR9024D Voltaire')
#0x0004 003 : (Switch portguid 0x0008f10400411c10: 'ISR9024D Voltaire')
# ...
#0x000d 004 : (Switch portguid 0x0008f10400401641: 'ISR2012 Voltaire sFB-2012')
#0x000e 004 : (Switch portguid 0x0008f10400401642: 'ISR2012 Voltaire sFB-2012')
#0x000f 004 : (Switch portguid 0x0008f10400401643: 'ISR2012 Voltaire sFB-2012')
#0x0011 004 : (Switch portguid 0x0008f10400411d48: 'ISR9024D Voltaire')
#0x0012 004 : (Switch portguid 0x0008f104003f1e54: 'ISR2012/ISR2004 Voltaire sLB-2024')
#0x0013 004 : (Switch portguid 0x0008f104003f1e55: 'ISR2012/ISR2004 Voltaire sLB-2024')
# ...
#0x0088 007 : (Channel Adapter portguid 0x0002c9020024d26d: 'minos514')
#0x008c 005 : (Channel Adapter portguid 0x0002c9020024c961: 'minos311')
#0x0090 012 : (Channel Adapter portguid 0x0002c90200246a89: 'minos362')
# ...
#0x0404 001 : (Channel Adapter portguid 0x0002c9020024d3e1: 'minos703')
#0x0405 002 : (path #2 out of 4: portguid 0x0002c9020024d3e1)
#0x0406 003 : (path #3 out of 4: portguid 0x0002c9020024d3e1)
#0x0407 004 : (path #4 out of 4: portguid 0x0002c9020024d3e1)
#969 valid lids dumped
sub read_routing_file
{
	my $regenerate_cache = $_[0];
	my $ca_name          = $_[1];
	my $ca_port          = $_[2];

	my $cache_file = get_routing_cache_file($ca_name, $ca_port);

	if ($regenerate_cache || !(-f "$cache_file")) {
		generate_routing_file($ca_name, $ca_port);
	}

	my %guid2name = undef;

	my ($swlid, $swguid);
	my $line = "";
	open(IN, $cache_file);
	while ($line = <IN>) {
		chomp $line;
		if (not $line) { next; }    # skip blank lines
		if ($line =~ /^Unicast lids/)    # starting table for new switch
		{
			($swlid, $swguid) = ($line =~ /Lid (\w+) guid (0x\w+) /);
			next;
		}
		if ($line =~ /^(0x\w+) (\d+) : .* portguid (0x\w+): '(\w+)'/) {
			my ($lid, $port, $guid, $name) = ($1, $2, $3, $4);
			$routing{$swguid}{$lid} = $port;
			$guid2name{$guid} = $name;
			$name2lids{$name} = ($name2lids{$name}, $lid);
		} elsif ($line =~ /^(0x\w+) (\d+) : .* portguid (0x\w+)/) {
			my ($lid, $port, $guid) = ($1, $2, $3);
			my $name = $guid2name{$guid};
			$routing{$swguid}{$lid} = $port;
			$name2lids{$name} = ($name2lids{$name}, $lid);
		}
	}
	close(IN);
}

# Get routing file cached
sub get_routing_cache_file
{
	my $ca_name = $_[0];
	my $ca_port = $_[1];
	ensure_cache_dir;
	return ("$IBswcountlimits::cache_dir/lft_dump-$ca_name-$ca_port");
}

# generate the routing file
sub generate_routing_file
{
	my $ca_name      = $_[0];
	my $ca_port      = $_[1];
	my $cache_file   = get_routing_cache_file($ca_name, $ca_port);
	my $extra_params = get_ca_name_port_param_string($ca_name, $ca_port);

	`dump_lfts.sh $extra_params > $cache_file`;
	if ($? != 0) {
		die "Execution of ibnetdiscover failed with errors\n";
	}
}

File: src/ibgraphfabric.c
/*
 * Copyright (c) 2004-2009 Voltaire Inc.  All rights reserved.
 * Copyright (c) 2007 Xsigo Systems Inc.  All rights reserved.
 * Copyright (c) 2008 Lawrence Livermore National Lab.  All rights reserved.
 * Copyright (c) 2009 HNR Consulting.  All rights reserved.
 * Copyright (c) 2010 Lawrence Livermore National Lab.  All rights reserved.
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the
 * OpenIB.org BSD license below:
 *
 *     Redistribution and use in source and binary forms, with or
 *     without modification, are permitted provided that the following
 *     conditions are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer in the documentation and/or other materials
 *        provided with the distribution.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

#if HAVE_CONFIG_H
#  include <config.h>
#endif				/* HAVE_CONFIG_H */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdarg.h>
#include <time.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <inttypes.h>
#include <regex.h>

#include <complib/cl_nodenamemap.h>
#include <infiniband/ibnetdisc.h>
#include <infiniband/mad.h>

char *argv0 = NULL;

static char *node_name_map_file = NULL;
static nn_map_t *node_name_map = NULL;

struct ibmad_port *ibmad_port;
ib_portid_t *ibd_sm_id;
ib_portid_t sm_portid = { 0 };

char *node_guid_str = NULL;
char *root_regex = NULL;
char *ignore_regex = NULL;
char *cluster_name = NULL;
uint64_t node_guid = 0;
char *dr_path = NULL;
int combine_edges = 0;
char *fromname = NULL;
char *toname = NULL;

ibnd_node_t *fromnode = NULL;
ibnd_node_t *tonode = NULL;

#define HTSZ 137
#define HASHGUID(guid) (((uint32_t)(((uint32_t)(guid) * 101) ^ ((uint32_t)((guid) >> 32) * 103))) % HTSZ)
struct lnode {
	struct lnode *next;
	ibnd_node_t *node;
	char port_visited[256];
};

struct local_fabric {
	int num_nodes;
	int rank_sep;
	struct lnode *nodes[HTSZ];
	struct lnode *roots;
	struct lnode *leafs;
} lfab;

static struct lnode *find_lnode(uint64_t guid)
{
	int hash = HASHGUID(guid) % HTSZ;
	struct lnode *lnode;

	for (lnode = lfab.nodes[hash]; lnode; lnode = lnode->next)
		if (lnode->node->guid == guid)
			return lnode;
	return (NULL);
}
static void insert_node(struct lnode *lnode)
{
	int hash_idx = HASHGUID(lnode->node->guid) % HTSZ;
	/* keep track by hash */
	lnode->next = lfab.nodes[hash_idx];
	lfab.nodes[hash_idx] = lnode;
}

static void flag_port(ibnd_node_t *node, uint8_t portnum)
{
	struct lnode *lnode = find_lnode(node->guid);
	if (!lnode) {
		lnode = calloc(1, sizeof(*lnode));
		lnode->node = node;
		insert_node(lnode);
	}
	lnode->port_visited[portnum] = 1;
}

static int port_flagged(uint64_t guid, uint8_t portnum)
{
	struct lnode *node = find_lnode(guid);
	if (!node)
		return 0;
	return (node->port_visited[portnum]);
}

static void id_root(ibnd_node_t *node, char *node_name)
{
	int rc;
	static regex_t exp;
	static int regex_compiled = 0;
	static int regex_failed = 0;

	if (!root_regex || regex_failed)
		return;

	if (!regex_compiled) { /* only compile it one time */
		if ((rc = regcomp(&exp, root_regex, REG_ICASE |
				REG_NOSUB | REG_EXTENDED)) != 0) {
			fprintf(stderr, "ERROR: regcomp failed on \"%s\": %d\n",
				root_regex, rc);
			regex_failed = 1;
			return;
		}
		regex_compiled = 1;
	}

	if (regexec(&exp, node_name, 0, NULL, 0) == 0) {
		struct lnode *lnode = calloc(1, sizeof(*lnode));
		if (!lnode) {
			fprintf(stderr, "%s:%d calloc FAILED! (%s)\n",
				__FUNCTION__, __LINE__, strerror(errno));
			return;
		}
		lnode->node = node;
		if (lfab.roots)
			lnode->next = lfab.roots;
		lfab.roots = lnode;
	}
}

static void add_to_leafs(ibnd_node_t *node)
{
	struct lnode *lnode = calloc(1, sizeof(*lnode));
	if (!lnode) {
		fprintf(stderr, "%s:%d calloc FAILED! (%s)\n",
			__FUNCTION__, __LINE__, strerror(errno));
		return;
	}
	lnode->node = node;
	if (lfab.leafs)
		lnode->next = lfab.leafs;
	lfab.leafs = lnode;
}

static int ignore_node(char *node_name)
{
	int rc;
	static regex_t exp;
	static int regex_compiled = 0;
	static int regex_failed = 0;

	if (!ignore_regex || regex_failed)
		return 0;

	if (!regex_compiled) { /* only compile it one time */
		if ((rc = regcomp(&exp, ignore_regex, REG_ICASE |
				REG_NOSUB | REG_EXTENDED)) != 0) {
			fprintf(stderr, "ERROR: regcomp failed on \"%s\": %d\n",
				root_regex, rc);
			regex_failed = 1;
			return 0;
		}
		regex_compiled = 1;
	}

	return (regexec(&exp, node_name, 0, NULL, 0) == 0);
}

struct links {
	uint8_t pnum;
	uint64_t rem_guid;
	uint8_t rem_pnum;
	int combine_num;
};

static void process_node(ibnd_node_t *node, void *data)
{
	static regex_t exp;
	static int regex_compiled = 0;
	char re_str[256];

	int rc, pnum, i;
	char label[256];
	char attr[256];
	char *node_name = NULL;
	attr[0] = '\0';
	ibnd_port_t *rem_port = NULL;
	struct links links[256];
	int num_links = 0;

	node_name = remap_node_name(node_name_map,
					node->guid,
					node->nodedesc);

	if (ignore_node(node_name))
		return;

	id_root(node, node_name);

	if (!regex_compiled && cluster_name) { /* only compile it one time */
		snprintf(re_str, 256, "^%s.*$", cluster_name);
		re_str[255] = '\0';
		if ((rc = regcomp(&exp, re_str, REG_ICASE |
				REG_NOSUB | REG_EXTENDED)) != 0) {
			fprintf(stderr, "ERROR: regcomp failed on \"%s\": %d\n",
				re_str, rc);
		}
		regex_compiled = 1;
	}

	if (regex_compiled && cluster_name
	    && (regexec(&exp, node_name, 0, NULL, 0) == 0)) {
		snprintf(label, 256, "%s", node_name);
		add_to_leafs(node);
	} else
		snprintf(label, 256, "%s\\nG: x%016lx", node_name, node->guid);

	label[255] = '\0';
	printf("   x%016lx [label=\"%s\"%s];\n", node->guid, label, attr);

	for (pnum = 1; pnum <= node->numports; pnum++) {
		if (!node->ports[pnum])
			continue;
		rem_port = node->ports[pnum]->remoteport;
		if (!rem_port)
			continue;
		if (!port_flagged(node->guid, pnum)) {
			char *rem_node_name = remap_node_name(node_name_map,
							rem_port->node->guid,
							rem_port->node->nodedesc);
			if (ignore_node(rem_node_name)) {
				free(rem_node_name);
				continue;
			}
			free(rem_node_name);

			flag_port(rem_port->node, rem_port->portnum);
			if (combine_edges) {
				int f = 0;
				for (i = 0; i < num_links; i++)
					if (links[i].rem_guid == rem_port->node->guid) {
						links[i].combine_num++;
						f = 1;
						break;
					}
				if (f)
					continue;
			}
			links[num_links].pnum = pnum;
			links[num_links].rem_guid = rem_port->node->guid;
			links[num_links].rem_pnum = rem_port->portnum;
			links[num_links].combine_num = 1;
			num_links++;
		}
	}

	for (i = 0; i < num_links; i++) {
		if (links[i].combine_num > 1) {
			printf("   x%016lx -> x%016lx [label=\"%d\", "
				"arrowhead=\"none\", color=\"blue\"];\n",
				node->guid, links[i].rem_guid,
				links[i].combine_num);
		} else {
			printf("   x%016lx -> x%016lx "
				"[taillabel=\"%d\", headlabel=\"%d\", "
				"arrowhead=\"none\"];\n",
				node->guid, links[i].rem_guid,
				links[i].pnum, links[i].rem_pnum);
		}
	}

	free(node_name);
}

static void print_dot_file(ibnd_fabric_t *fabric)
{
	ibnd_node_t *node = fabric->nodes;
	while (node) {
		lfab.num_nodes++;
		node = node->next;
	}
	lfab.rank_sep = 1 + ((lfab.num_nodes) * 0.5/10);
	printf("digraph G {\n");
	printf("   node [shape=record, fontsize=9];\n");
	printf("   graph [outputorder=nodesfirst, ranksep=\"%d equally\"];\n",
		lfab.rank_sep);

	ibnd_iter_nodes(fabric, process_node, NULL);

	if (lfab.roots) {
		struct lnode *lnode;
		printf("{rank=min; ");
		for (lnode = lfab.roots; lnode; lnode = lnode->next) {
			printf("x%016lx; \n", lnode->node->guid);
		}
		printf("}\n");
	}

	if (lfab.leafs) {
		struct lnode *lnode;
		printf("{rank=max; ");
		for (lnode = lfab.leafs; lnode; lnode = lnode->next) {
			printf("x%016lx; \n", lnode->node->guid);
		}
		printf("}\n");
	}

	printf("}\n");
}

/** =========================================================================
*/
static ibnd_node_t *find_node(ibnd_fabric_t *fabric, char *name)
{
	ibnd_node_t node;
	for (node = fabric->nodes; node; node->next) {
		if (strcmp(node->nodedesc, name) == 0) {
			return (node);
		}
	}
	return (NULL);
}

/** =========================================================================
 */
static int
usage(void)
{
        fprintf(stderr,
"%s [options]\n"
"Usage: generate a slurm topology file to be fed to slurm for better node utilization\n"
"\n"
"Options:\n"
"  -S <guid> generate for the node specified by the port guid\n"
"  -G <guid> Same as \"-S\" for compatibility with other diags\n"
"  -D <dr_path> generate for the node specified by the DR path given\n"
"  --hops -n <hops> hops to progress away from node given in -G,-S, or -D\n"
"                   default == 0\n"
"  -e combine edges between nodes into one edge with an associated weight\n"
"  -g <cluster_name> DON'T print GUID's on nodes if labeled <cluster_name>N\n"
"  -r <regex> regex to identify root switches\n"
"  -i <regex> regex to identify nodes to ignore\n"
"  -M <node1:node2> plot route(s) from node1 to node2 in red\n"

"  --node-name-map <map> specify alternate node name map\n"
"  --Ca, -C <ca>         Ca name to use\n"
"  --Port, -P <port>     Ca port number to use\n"
"  --timeout, -t <ms>    timeout in ms\n"
"  --verbose, -v         increase verbosity level\n"
"  -R GNDN.  This option is here for backward compatibility\n"
"\n"
, argv0
);
        return (0);
}

int main(int argc, char **argv)
{
	int resolved = -1;
	ib_portid_t portid = { 0 };
	int rc = 0;
	ibnd_fabric_t *fabric = NULL;
	struct ibnd_config config = { 0 };

	int mgmt_classes[4] = { IB_SMI_CLASS, IB_SMI_DIRECT_CLASS, IB_SA_CLASS,
		IB_PERFORMANCE_CLASS
	};
	char *ibd_ca = NULL;
	int ibd_ca_port = 0;
	int ibd_timeout = 200;

        char  ch = 0;
        static char const str_opts[] = "hS:G:D:r:ei:g:n:C:P:t:vM:";
        static const struct option long_opts [] = {
           {"help", 0, 0, 'h'},
	   {"node-name-map", 1, 0, 1},
	   {"hops", 1, 0, 'n'},
	   {"Ca", 1, 0, 'C'},
	   {"Port", 1, 0, 'P'},
	   {"timeout", 1, 0, 't'},
	   {"verbose", 1, 0, 'v'},
	   {0, 0, 0, 0}
        };

	argv0 = argv[0];

        while ((ch = getopt_long(argc, argv, str_opts, long_opts, NULL))
                != -1)
        {
                switch (ch)
                {
			case 'S':
			case 'G':
				node_guid_str = strdup(optarg);
				break;
			case 'D':
				dr_path = strdup(optarg);
				break;
			case 'r':
				root_regex = strdup(optarg);
				break;
			case 'e':
				combine_edges = 1;
				break;
			case 'M':
			{
				char *tmp = strdup(optarg);
				fromname = strdup(strtok(tmp, ":"));
				toname = strdup(strtok(NULL, ":"));
				free(tmp);
				break;
			}
			case 'i':
				ignore_regex = strdup(optarg);
				break;
			case 'g':
				cluster_name = strdup(optarg);
				break;
			case 'n':
				config.max_hops = (int)strtol(optarg, NULL, 0);
				break;
			case 1:
				node_name_map_file = strdup(optarg);
				break;
			case 'C':
				ibd_ca = strdup(optarg);
				break;
			case 'P':
				ibd_ca_port = atoi(optarg);
				break;
			case 't':
				ibd_timeout = atoi(optarg);
				break;
			case 's':
				/* srcport is not required when resolving via IB_DEST_LID */
				if (ib_resolve_portid_str_via(&sm_portid, optarg, IB_DEST_LID,
							      0, NULL) < 0)
					fprintf(stderr, "cannot resolve SM destination port %s",
						optarg);
				ibd_sm_id = &sm_portid;
				break;
			case 'R': /* noop */
				break;
                        case 'h':
                        default:
                        	exit(usage());
                }
	}

	ibmad_port = mad_rpc_open_port(ibd_ca, ibd_ca_port, mgmt_classes, 4);
	if (!ibmad_port) {
		fprintf(stderr, "Failed to open port; %s:%d\n", ibd_ca, ibd_ca_port);
		exit(-1);
	}

	if (ibd_timeout) {
		mad_rpc_set_timeout(ibmad_port, ibd_timeout);
		config.timeout_ms = ibd_timeout;
	}

	node_name_map = open_node_name_map(node_name_map_file);

	/* limit the scan to around the target */
	if (dr_path) {
		if ((resolved =
		     ib_resolve_portid_str_via(&portid, dr_path, IB_DEST_DRPATH,
					       NULL, ibmad_port)) < 0) {
			fprintf(stderr, "Failed to resolve %s", dr_path);
			rc = -1;
			goto close_port;
		}
	} else if (node_guid_str) {
		if ((resolved =
		     ib_resolve_portid_str_via(&portid, node_guid_str,
					       IB_DEST_GUID, ibd_sm_id,
					       ibmad_port)) < 0) {
			fprintf(stderr, "Failed to resolve %s", node_guid_str);
			rc = -1;
			goto close_port;
		}
	}

	if (resolved >= 0) {
		if (!config.max_hops)
			config.max_hops = 1;
		fabric = ibnd_discover_fabric(ibd_ca, ibd_ca_port, &portid, &config);
	} else {
		fabric = ibnd_discover_fabric(ibd_ca, ibd_ca_port, NULL, &config);
	}

	if (!fabric) {
		fprintf(stderr, "ibnd_discover_fabric failed\n");
		rc = -1;
		goto close_port;
	}

	if (fromname && toname) {
		fprintf(stderr, "Plotting from %s => to %s\n",
			fromname, toname);
		fromnode = find_node(fabric, fromname);
		tonode = find_node(fabric, toname);
		if (!fromnode || !tonode) {
			if (!fromnode)
				fprintf(stderr,
					"Error: %s not found in fabric\n",
					fromname);
			if (!tonode)
				fprintf(stderr,
					"Error: %s not found in fabric\n",
					toname);
			free(fromname);
			free(toname);
			goto destroy_fabric;
		}
		free(fromname);
		free(toname);
	}

	print_dot_file(fabric);

destroy_fabric:
	ibnd_destroy_fabric(fabric);

close_port:
	mad_rpc_close_port(ibmad_port);
	close_node_name_map(node_name_map);
	exit(rc);
}
